### PR TITLE
[Fixes BIG bug] register blade directives after resolving blade compiler

### LINKED ISSUE #2038: @role displaying on page when spatie/markdown in use
**Describe the bug**
When spatie/laravel-markdown and spatie/laravel-permission are installed the @role type directives are displayed as part of the HTML.

**Versions**
- spatie/laravel-permission package version: 5.5.0
- spatie/laravel-markdown: 2.2.1
- illuminate/framework package: v8.83.2

PHP version: 8.0.16

Database version: mysql  v 8.0.28

**To Reproduce**
Have both laravel-markdown and laravel-permission installed and inside a blade view include both <x-markdown> and an @role type directive. When the page is rendered the @role is output as text

Here is my example code and/or tests showing the problem in my app:
`<x-markdown>{{ $markdown }}</x-markdown> @role('admin')<a href='/admin'>Admin Panel</a>@endrole`

**Example Application**
Here is a link to my Github repo containing a minimal Laravel application which shows my problem: https://github.com/kaptk2/permission-issue

**Expected behavior**
I expected the custom blade directives to be honored and not displayed to the end user.

 **Additional context**
This issue also exists with Laravel 9, the example repository above is using Laravel 9.

**Environment (please complete the following information, because it helps us investigate better):**
 - OS: Ubuntu
 - Version 20.04

### PR DESCRIPTION
**Pull request by tabacitu**
This is a cleaner alternative to https://github.com/spatie/laravel-permission/pull/2047

Fixes #2038, #2044, #2045

### CODE REVIEW

**Comment by erikn69** [comment]
Looks great 👍

**Comment by erikn69** [comment]
@freekmurze after https://github.com/spatie/laravel-markdown/pull/35 this could be merged if it's better for everyone, 
it would revert to 5.5.0 functionality

**Comment by tabacitu** [comment]
Thanks for the double-check @erikn69 🎉

TLDR for @freekmurze : 
- this package was using `afterResolving('blade.compiler')` until 6 days ago;
- PR #2039 changed that; the change made this package the final stop for registering blade directives; when other packages tried to register blade directives after it... they wouldn't work, they'd be output as plain text;
- this PR reverts #2039 , and leaves the campsite a little cleaner than before;

For details, see #2047 . But it's a lot.

Let me know if you have questions @freekmurze .
Cheers!

**Comment by erikn69** [comment]
For the records, seems like `Blade::directive()`/`$this->app['blade.compiler']` force to resolve `BladeCompiler` inmediatly,  after that others `afterResolving` callbacks never gets executed because `BladeCompiler` is already resolved  (race condition)

**Comment by erikn69** [comment]
For the records, seems like `Blade::directive()`/`$this->app['blade.compiler']` force to resolve `BladeCompiler` inmediatly,  after that others `afterResolving` callbacks never gets executed because `BladeCompiler` is already resolved  (race condition)

**Comment by erikn69** [comment]
https://github.com/laravel/framework/pull/41397#issuecomment-1062169244

**Comment by fredsal** [comment]
@freekmurze ping :/

**Comment by erikn69** [comment]
>don't register Blade extensions in the `register` method.

https://github.com/laravel/framework/pull/41397#issuecomment-1062223510

**Comment by tabacitu** [comment]
While I agree that a fix upstream would be nice, I still think this should be merged ASAP. Every day we delay, this package has 26.000 downloads with potentially breaking setups.

Please don't drag this on any longer.

@freekmurze ?

**Comment by pxpm** [comment]
I am a bit tired to do the Pr today. I can do it tomorrow morning. Don't worry, by lunch time everything is solved!

**Comment by erikn69** [comment]
>@tabacitu the update is: use $this->callAfterResolving instead of $this->app->afterResolving

Still this PR needs to be merged

**Comment by tabacitu** [comment]
Agreed. This PR still needs to be merged.

**Comment by tabacitu** [comment]
I've just changed the PR to use the solution Taylor Otwell and Graham Campbell proposed in https://github.com/laravel/framework/pull/41397 - using `callAfterResolving()` instead of `afterResolving()`. This solves the problem for now. And for the future 💪

@erikn69 can you please double-check so that it's easier for @freekmurze to merge?

**Comment by pxpm** [comment]
@tabacitu @erikn69 @freekmurze 

I was the one discussing this with Taylor Otwell and Graham Campbell in https://github.com/laravel/framework/pull/41397 . 

The solution proposed by them (and it does what we want), is using `$this->callAfterResolving()`, that is basically the same as calling `$this->app->afterResolving()` with an aditional parameter to run the callback if the class is already resolved. 

I don't quite agree with their solution for this problem, I was searching the Container api/docs like mad to find something like this, it happens it's not in `Container`, but in `ServiceProvider`. 

Using `$this->callAfterResolving()` 100% solves the problem. This should be the "de-facto" solution here, since this package was using `afterResolving()`, changing to `Blade::directive()` - same as `app('blade.compiler')` may have big implications in packages that rely on `afterResolving()` of this package. 

@freekmurze  please consider merging this PR as a Backward Compatible change. It will fix your package problem once and for all as Taylor suggested, but would not break so many other packages as changing to `Blade::directive()` approach. 

Wish you all guys the best,
Pedro

**Comment by freekmurze** [comment]
Thanks!

**Comment by tabacitu** [comment]
No, thank you! And thanks @erikn69 and everyone else for the double-check. 

Wheew... This was a nasty one 😅 Glad it's over.
